File: iphone/WinLock/Models/ApiResponse.swift
import Foundation

/// Standard response envelope shared by every WinLock API endpoint.
struct ApiError: Decodable {
    let code: String
    let message: String
}

struct ApiResponse<T: Decodable>: Decodable {
    let success: Bool
    let message: String?
    let error: ApiError?
    let data: T?
}

/// Placeholder used for endpoints that return no data payload.
struct EmptyData: Decodable {}

/// Errors surfaced from the WinLock service API.
enum APIError: LocalizedError {
    case invalidResponse
    case server(code: String, message: String)
    case invalidPairingPayload

    var errorDescription: String? {
        switch self {
        case .invalidResponse: return "Invalid response from the laptop"
        case .server(_, let message): return message
        case .invalidPairingPayload: return "The pairing payload is invalid"
        }
    }
}

File: windows/WinLock.Protocol/Models/ApiResponse.cs
using System.Text.Json.Serialization;

namespace WinLock.Protocol.Models;

/// <summary>Standard response envelope used by every API endpoint.</summary>
public class ApiResponse
{
    public bool Success { get; init; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Message { get; init; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ApiError? Error { get; init; }

    public static ApiResponse SuccessResult(string? message = null) =>
        new() { Success = true, Message = message };

    public static ApiResponse Failure(string code, string message) =>
        new() { Success = false, Error = new ApiError { Code = code, Message = message } };
}

/// <summary>Response envelope carrying typed payload data.</summary>
public sealed class ApiResponse<T> : ApiResponse
{
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public T? Data { get; init; }

    public static ApiResponse<T> Ok(T data, string? message = null) =>
        new() { Success = true, Message = message, Data = data };

    public static new ApiResponse<T> Failure(string code, string message) =>
        new() { Success = false, Error = new ApiError { Code = code, Message = message } };
}